pkg/proxy/iptables/proxier.go

```diff
@@ -618,41 +592,28 @@ func buildNewEndpointsMap(allEndpoints []*api.Endpoints, curMap proxyEndpointMap
 		return
 	}
 
-	// Update service health check.  We include entries from the current map,
-	// with zero-length value, to trigger the healthchecker to stop reporting
-	// health for that service.
-	//
-	// This whole mechanism may be over-designed.  It builds a list of endpoints
-	// per service, filters for local endpoints, builds a string that is the
-	// same as the name, and then passes each (name, list) pair over a channel.
-	//
-	// I am pretty sure that there's no way there can be more than one entry in
-	// the final list, and passing an empty list as a delete signal is weird.
-	// It could probably be simplified to a synchronous function call of a set
-	// of NamespacedNames.  I am not making that simplification at this time.
-	//
-	// ServicePortName includes the port name, which doesn't matter for
-	// healthchecks. It's possible that a single update both added and removed
-	// ports on the same IP, so we need to make sure that removals are counted,
-	// with additions overriding them.  Track all endpoints so we can find local
-	// ones.
-	epsBySvcName := map[types.NamespacedName][]*endpointsInfo{}
-	for svcPort := range curMap {
-		epsBySvcName[svcPort.NamespacedName] = nil
-	}
+	// accumulate local IPs per service, ignoring ports
+	localIPs := map[types.NamespacedName]sets.String{}
 	for svcPort := range newMap {
-		epsBySvcName[svcPort.NamespacedName] = append(epsBySvcName[svcPort.NamespacedName], newMap[svcPort]...)
-	}
-	for nsn, eps := range epsBySvcName {
-		// Use a set instead of a slice to provide deduplication
-		epSet := sets.NewString()
-		for _, ep := range eps {
+		for _, ep := range newMap[svcPort] {
 			if ep.isLocal {
-				// kube-proxy health check only needs local endpoints
-				epSet.Insert(fmt.Sprintf("%s/%s", nsn.Namespace, nsn.Name))
+				nsn := svcPort.NamespacedName
+				if localIPs[nsn] == nil {
+					localIPs[nsn] = sets.NewString()
+				}
+				ip := strings.Split(ep.endpoint, ":")[0] // just the IP part
+				localIPs[nsn].Insert(ip)
 			}
 		}
-		healthChecker.UpdateEndpoints(nsn, epSet)
+	}
+	// produce a count per service
+	localEndpointCounts := map[types.NamespacedName]int{}
+	for nsn, ips := range localIPs {
+		localEndpointCounts[nsn] = len(ips)
+	}
+	// update healthcheck endpoints
+	if err := healthChecker.SyncEndpoints(localEndpointCounts); err != nil {
```

Comment:
Would it be better to call `SyncEndpoints` in `OnEndpointsUpdate` instead of in `buildNewEndpointsMap` (just like what `OnServiceUpdate` does)?